File: 10/first.rb
def adapters(file)
  File.open(file, 'r').each_line.map { |line| line.strip.to_i }
end

sorted = adapters('input.sample.txt').sort
sorted = adapters('input.sample2.txt').sort
sorted = adapters('input.txt').sort

left = sorted.dup
right = sorted.dup

last = left.pop
right.shift

jump_map = left.zip(right).each_with_object(Hash.new { |_key| 1 }) do |(a, b), map|
  map[b-a] += 1
end


puts jump_map.inspect
puts jump_map[3] * jump_map[1]
puts last+3

File: 10/first.ex
adapters = fn file ->
  file
  |> File.stream!()
  |> Stream.map(&String.trim/1)
  |> Stream.map(&String.to_integer/1)
  |> Enum.sort()
end

sorted = adapters.('input.sample.txt')
sorted = adapters.('input.sample2.txt')
sorted = adapters.('input.txt')

sorted
|> Enum.zip(tl(sorted))
|> Enum.reduce(%{}, fn {a, b}, acc ->
  Map.update(acc, b-a, 2, fn count -> count + 1 end)
end)
|> IO.inspect()
|> (fn map -> map[3] * map[1] end).()
|> IO.inspect()
